DS-3572: Check authorization for a specified user instead of currentUser

tomdesair opened the pull request:
https://jira.duraspace.org/browse/DS-3572
This PR build on PR https://github.com/DSpace/DSpace/pull/1715 but fixes some tests.

pnbecker committed:
DS-3572: Check authorization for a specified user instead of currentUser

pnbecker committed:
DS-3572: Adding simple unit test for DS-3572.

Commit:
Restore GroupServiceImpl.isMember logic + fix tests

Commit:
Attempt to fix contstraint violation

Commit:
Fix DSpace AIP IT tests: Set correct membership for admin

tdonohue commented:
Looks good, and glad to see this passes!

Just as a note, this afternoon, I was trying to help cherry-pick #1694 to `master` and realized that the work in this 1715 PR is dependent on 1694 slightly (as 1694 changes some of the logic in `GroupServiceImpl.java`).  So, while it shouldn't hold up merging this PR into `dspace-6_x`, I suspect we will need to merge #1694 into `master` *before* this PR will cherry-pick easily.

As of yet, I've had difficulty myself porting #1694 to `master` and haven't been able to get it to pass unit tests.  @tomdesair , if you have time to chip in on this, or take it over, I've started a #1694 port to master in this branch (but it fails a lot of unit tests, so somewhere my conflict merging must've gone wrong): https://github.com/tdonohue/DSpace/tree/DS-3552-master

Commit:
Improve tests + make GroupService.isMember method more performant for special groups

Commit:
DS-3572: Fix bug where normal group membership is ignored if special groups are present + added tests

tdonohue commented:
Minor update to my previous note about #1694 port to master. I finally got it building properly and passing unit tests. See #1769. So, I'd recommend it is merged to master prior to any version of this PR (as the cherry pick should be much smoother that way).

tomdesair commented:
@pnbecker I've fixed the tests and I've implemented your suggestion to make the `isMember` call with special groups more performant. I also updated the `GroupTest` class to provide better test coverage of the `isMember` methods.

pnbecker commented:
This should be cherry-picked to merge as soon as it get merged to dspace-6_x.

pnbecker commented on dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java:
```diff
@@ -152,59 +157,80 @@ public boolean isDirectMember(Group group, EPerson ePerson) {
-    public boolean isMember(Group owningGroup, Group childGroup) {
-        return owningGroup.contains(childGroup);
+    public boolean isMember(Context context, Group owningGroup, Group childGroup) throws SQLException {
+        return group2GroupCacheDAO.findByParentAndChild(context, owningGroup, childGroup) != null;
```
Please take a look into the javadoc in GroupService.isMember. This method should explicitly check only direct members. Using group2GroupCache is too much here. But we should add another method doing this.

pnbecker reviewed (commented):
I contacted @tomdesair in slack. My request should be easy to change.

This PR contains commits I wrote myself. It should be reviewed by at least one another committer. Nevertheless I'm +1 as soon as my request regarding the method `GroupService.isMember(Group, Group)` is resolved.

Commit:
DS-3572: Restored behaviour of GroupService.isMember and moved new behaviour to GroupService.isParentOf

Commit:
DS-3572: Renamed epersonInGroup to isEPersonInGroup

pnbecker reviewed (approved):
As mentioned before two commits are written by myself. So another approval would be good.

+1

wwelling commented:
Deployed and tested the following in the UI:
- register user
- create admin *dspace command
- log in/log out as user and admin
- create community as admin
- create collection as admin
- assigned user as submitter
- submitted to collection as user
- submitted to collection as admin
- edit item metadata as user
- edit item metadata as admin
- removed anonymous access to bitstream, user must now request a copy
- update profile as user
- update profile as admin
- versioned item as admin
- add user as administrator policy on a sub-community
- user able to create collection in sub-community

+1

tomdesair commented:
Thanks for testing @wwelling !

hardyoyo reviewed (approved):
Looks right, passes tests, is submitted by a trusted dev, and tested by another trusted dev. +1 I approve.

pnbecker commented:
Cherry-picked to master.
